core/src/main/java/org/elasticsearch/action/bulk/BulkRequestHandler.java

```diff
@@ -0,0 +1,160 @@
+                BulkResponse bulkResponse = Retry
+                        .on(EsRejectedExecutionException.class)
+                        .policy(backoffPolicy)
+                        .withSyncBackoff(client, bulkRequest);
+                afterCalled = true;
+                listener.afterBulk(executionId, bulkRequest, bulkResponse);
+            } catch (Exception e) {
+                if (!afterCalled) {
+                    logger.warn("Failed to executed bulk request {}.", e, executionId);
+                    listener.afterBulk(executionId, bulkRequest, e);
+                }
+            }
+        }
+
+        @Override
+        public boolean awaitClose(long timeout, TimeUnit unit) throws InterruptedException {
+            // we are "closed" immediately as there is no request in flight
+            return true;
+        }
+    }
+
+    private static class AsyncBulkRequestHandler extends BulkRequestHandler {
+        private final BackoffPolicy backoffPolicy;
+        private final BulkProcessor.Listener listener;
+        private final Semaphore semaphore;
+        private final int concurrentRequests;
+
+        private AsyncBulkRequestHandler(Client client, BackoffPolicy backoffPolicy, BulkProcessor.Listener listener, int concurrentRequests) {
+            super(client);
+            this.backoffPolicy = backoffPolicy;
+            assert concurrentRequests > 0;
+            this.listener = listener;
+            this.concurrentRequests = concurrentRequests;
+            this.semaphore = new Semaphore(concurrentRequests);
+        }
+
+        @Override
+        public void execute(BulkRequest bulkRequest, long executionId) {
+            boolean success = false;
+            boolean acquired = false;
+            try {
+                listener.beforeBulk(executionId, bulkRequest);
+                semaphore.acquire();
+                acquired = true;
+                Retry.on(EsRejectedExecutionException.class)
+                        .policy(backoffPolicy)
+                        .withAsyncBackoff(client, bulkRequest, new ActionListener<BulkResponse>() {
+                            @Override
+                            public void onResponse(BulkResponse response) {
+                                try {
+                                    listener.afterBulk(executionId, bulkRequest, response);
+                                } finally {
+                                    semaphore.release();
+                                }
+                            }
+
+                            @Override
+                            public void onFailure(Throwable e) {
+                                try {
+                                    listener.afterBulk(executionId, bulkRequest, e);
+                                } finally {
+                                    semaphore.release();
+                                }
+                            }
+                        });
+                success = true;
```

Comment:
Actually, it is required to not release the semaphore until after we're completely done. Otherwise, we might break the limit of maximum allowed concurrent requests the user has defined. We could maybe release the semaphore before we delegate to the listener (line 121 and line 131). Consider this execution trace:

maximum allowed number of concurrent requests: 2

1. A bulk request B0 comes in, we acquire the semaphore. Allowed concurrent requests = 1
2. we issue the request but hit a `EsRejectedExecutionException`. We back off for - say - 50 ms
3. A second bulk request B1 comes in, we acquire the semaphore. Allowed concurrent requests = 0
4. Another bulk request B2 comes in, we acquire the semaphore and we are now blocked as allowed concurrent requests = 0.
5. The first retry request on B0 is sent, returns and we're done. We call `onSuccess` on the callback listener in `AsyncBulkRequestHandler`which in turn will notify its delegate listener. Now we release the semaphore. Post condition: Number of allowed concurrent requests = 1.
6. B2 can now continue.

If we get rid of the `success` check we would release the semaphore immediately after the first attempt to send a bulk request and B2 would break the limit on the number of maximum allowed concurrent requests in step 4 in the above scenario.

@nik9000: Does this answer your remark?